DCHdigi/src/DCHdigi.cpp

```diff
@@ -0,0 +1,671 @@
+        oDCHsimdigi_association.setDigi( oDCHdigihit  );
+        oDCHsimdigi_association.setSim( input_sim_hit  );
+
+        }// end loop over hit collection
+
+    /////////////////////////////////////////////////////////////////
+    return std::make_tuple<colltype_out,colltype_out2>(std::move(output_digi_hits),std::move(output_digi_sim_association));
+}
+
+///////////////////////////////////////////////////////////////////////////////////////
+///////////////////////       finalize       //////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////////////
+StatusCode DCHdigi::finalize()
+{
+    if( m_create_debug_histos.value() )
+    {
+        std::unique_ptr<TFile> ofile{TFile::Open ( "dch_digi_alg_debug.root", "recreate" ) };
+        ofile->cd();
+        hDpw->Write();
+        hDww->Write();
+        hSxy->Write();
+        hSz->Write();
+    }
+
+    return StatusCode::SUCCESS;
+}
+
+///////////////////////////////////////////////////////////////////////////////////////
+///////////////////////       ThrowException       ////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////////////
+void DCHdigi::ThrowException(std::string s) const {
+    error() << s.c_str()  << endmsg;
+    throw std::runtime_error(s);
+  }
+
+///////////////////////////////////////////////////////////////////////////////////////
+///////////////////////       PrintConfiguration       ////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////////////
+void DCHdigi::PrintConfiguration(std::ostream& io)
+{
+    io << "DCHdigi will use the following components:\n";
+    io << "\tGeometry Service: "                  << m_geoSvcName.value().c_str()           << "\n";
+    io << "\tUID Service: "                       << m_uidSvcName.value().c_str()           << "\n";
+    io << "\tDetector name: "                     << m_DCH_name.value().c_str()             << "\n";
+    io << "\t\t|--Volume bitfield: "              << m_decoder->fieldDescription().c_str()  << "\n";
+    io << "\t\t|--Number of layers: "             << dch_data->database.size()              << "\n";
+    io << "\tCluster distributions taken from: "  << m_fileDataAlg.value().c_str()          << "\n";
+    io << "\tResolution along the wire (mm): "    << m_z_resolution.value()                 << "\n";
+    io << "\tResolution perp. to the wire (mm): " << m_xy_resolution.value()                << "\n";
+    return;
+}
+
+void DCHdigi::PrepareRandomEngine(const edm4hep::EventHeaderCollection&  headers) const
+{
+    uint32_t evt_n = headers[0].getEventNumber();
+    uint32_t run_n = headers[0].getRunNumber();
+    size_t seed = m_uidSvc->getUniqueID(evt_n, run_n, this->name() );
+    m_engine.seed(seed);
+    // test random engine...
+    m_engine.discard(10);
+    myRandom.SetSeed(seed+42);
```

Comment:
the mersenne twister may lead to similar results for consecutive seeds (that is, the digitizer might introduce correlations between events). Advancing the internal state some positions should mitigate that issue. I learnt this from @tmadlener, do you know if this is worth it?

The case of ROOT TRandom3 object, they do not provide the method to advance the internal state, so I put the +42 to the seed as a reminder. TRandom3 uses mersenne twister under the hood, so it may be affected by the same problems.

I will crosscheck with Geant4 experts, they had faced the same problem when they moved to MT for sure.

Reply:
fixed!